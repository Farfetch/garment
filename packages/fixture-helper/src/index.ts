export * from './executeRunner';
export * from './initFixtureHelper';
export * from './resolver';
import normalizePath = require('normalize-path');
import isValidPath = require('is-valid-path');

const containsValidPath = (value: string): boolean =>
  value.split(/\s/).some((s) => isValidPath(s));

export const normalizeContainedPath = (value: string): string => {
  return value
    .split(/(\s)/)
    .map((s) => {
      if (isValidPath(s)) {
        return normalizePath(s);
      } else {
        return s;
      }
    })
    .join('');
};

export function replaceTestPath<T>(
  obj: T,
  testPath: string,
  replaceWith = '/test_path'
): T {
  const normalizedTestPathRegEx = new RegExp(normalizePath(testPath), 'g');
  return JSON.parse(
    JSON.stringify(obj, (_, value) =>
      typeof value === 'string' && containsValidPath(value)
        ? normalizeContainedPath(value).replace(
            normalizedTestPathRegEx,
            replaceWith
          )
        : value
    )
  );
}
