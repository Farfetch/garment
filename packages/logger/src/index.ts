import chalk from 'chalk';
import stripAnsi from 'strip-ansi';

const levelsMap = {
  silly: 0,
  debug: 1,
  info: 2,
  success: 3,
  warn: 4,
  error: 5,
  silent: 6,
};

export type Level = keyof typeof levelsMap;

const logTypeToTextMap: { [key in Level]?: string } = {
  silly: chalk.keyword('purple')('silly'),
  debug: chalk.keyword('cyan')('debug'),
  info: chalk.keyword('lightblue')('info'),
  warn: chalk.keyword('yellow')('warning'),
  error: chalk.keyword('darkred')('error'),
  success: chalk.keyword('lime')('success'),
};

export function shouldPrint(level: Level, currentLevel: Level) {
  return levelsMap[level] >= levelsMap[currentLevel];
}

export const levels = Object.keys(levelsMap);

export class Logger {
  public static print(
    options: {
      prefix?: string;
      scope?: string;
      level?: Level;
      scopeColor?: string;
      dontLog?: boolean;
    } = {},
    ...args: any[]
  ) {
    const meta: string[] = [];
    const { prefix, scope, level, scopeColor = 'magenta', dontLog } = options;
    if (prefix) {
      meta.push(chalk.grey(prefix));
    }
    if (scope) {
      meta.push(chalk.keyword(scopeColor)(scope));
    }
    if (level) {
      meta.push(logTypeToTextMap[level] || level);
    }
    const joinedMeta = meta.join(' ');

    const indentedArgs = args
      .map((arg) => (Object(arg) !== arg ? arg : JSON.stringify(arg)))
      .join(' ')
      .split(/\n/)
      .join('\n' + ' '.repeat(stripAnsi(joinedMeta).length + 1));

    const consoleLogArgs: string[] = [joinedMeta, indentedArgs];
    if (!dontLog) {
      console.log(...consoleLogArgs);
    }
    return consoleLogArgs;
  }

  interceptors: ((entry: { level: Level; args: any[] }) => any)[] = [];

  constructor(
    public readonly prefix: string | undefined,
    public readonly name: string,
    public level: Level = 'info',
    public readonly color?: string
  ) {}

  public silly(...args: any[]) {
    this.display('silly', ...args);
  }

  public debug(...args: any[]) {
    this.display('debug', ...args);
  }

  public info(...args: any[]) {
    this.display('info', ...args);
  }

  public warn(...args: any[]) {
    this.display('warn', ...args);
  }

  public error(...args: any[]) {
    this.display('error', chalk.keyword('red')(...args));
  }

  public success(...args: any[]) {
    this.display('success', ...args);
  }

  public log(...args: any[]) {
    Logger.print({}, ...args);
  }

  public createChild(scope: string, color?: string) {
    return new Logger(this.prefix, scope, this.level, color);
  }

  public setLevel(level: Level) {
    this.level = level;
  }

  private display(level: Level, ...args: any[]) {
    if (
      !this.interceptors.every((fn) => {
        const result = fn({ level, args });
        return typeof result === 'boolean' ? result : true;
      })
    ) {
      return;
    }
    if (shouldPrint(level, this.level)) {
      Logger.print(
        {
          prefix: this.prefix,
          scope: this.name,
          level,
          scopeColor: this.color,
        },
        ...args
      );
    }
  }
}

export default new Logger('', '', 'info');
