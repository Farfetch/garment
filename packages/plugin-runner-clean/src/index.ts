import { defineRunner } from '@garment/runner';
import del from 'del';

/**
 * Deletes files and folders using glob patterns.
 */
export interface CleanRunnerOptions {}

export default defineRunner(async function (ctx) {
  ctx.logger.info(`Removing files`);

  ctx.input.forEach(async ({ path, baseDir }) => {
    if (baseDir) {
      await del(path, {
        cwd: baseDir,
      });
    } else {
      throw new Error(`Unable to delete "${path}". baseDir is empty`);
    }
  });

  ctx.logger.success(`Files removed`);
});
