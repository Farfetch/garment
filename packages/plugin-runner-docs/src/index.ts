import { defineRunner, defineOptions } from '@garment/runner';
import * as Path from 'path';
import { Project, SourceFile } from 'ts-morph';

export default defineRunner(
  defineOptions({
    template: ['path', 'Path to the template'],
    filename: ['string?', 'Name of the output file', 'README.md']
  }),
  async ctx => {
    const { template, filename } = ctx.options;
    const packageName = ctx.project.nodePackage?.name;

    const templateContent = ctx.renderTemplate(
      ctx.fs.readFileSync(template, 'utf8'),
      { packageName }
    );

    const templateLines = templateContent.split('\n');
    const updatedTemplateLines = [];

    const project = new Project({});
    const sourceFiles: Record<string, SourceFile> = {};

    let ignoreLine = false;

    for (const line of templateLines) {
      const match = /<!--\s*(.*)\s*-->/.exec(line);
      if (match) {
        const [op, ...args] = match[1].split(/\s+/);

        if (op === 'api') {
          const [filename, declarationName] = args;

          const sourceFile = getSourceByFilename(filename);

          const declaration =
            declarationName === 'firstInterface'
              ? sourceFile.getInterfaces()[0]
              : sourceFile.getInterfaceOrThrow(declarationName);

          const allResults = [];

          for (const member of declaration.getMembers()) {
            let result = [];
            const name = (member as any).getName();
            let type = member.getType().getText();
            let defaultValue;
            let comment;

            const [jsdoc] = member.getJsDocs();
            if (jsdoc) {
              comment = jsdoc.getDescription().trim();
              const tags: Record<string, string> = {};
              for (const tag of jsdoc.getTags()) {
                tags[tag.getTagName()] = tag.getComment()?.trim() ?? '';
              }

              if (tags.format === 'path') {
                type = 'path';
              }

              defaultValue = tags.defaultDescription || tags.default;
            }

            result.push(`\`${name}: ${type}\``);
            if (defaultValue) {
              result.push(`**Default:** \`${defaultValue}\``);
            }
            if (comment) {
              result.push(comment);
            }

            allResults.push(result.join('\n\n'));
          }

          ignoreLine = true;

          updatedTemplateLines.push(line);

          allResults.forEach((result, index) => {
            updatedTemplateLines.push(result);
            if (index === allResults.length - 1) {
              updatedTemplateLines.push('\n');
            } else {
              updatedTemplateLines.push('\n---\n');
            }
          });
        } else if (op === 'example' || op === 'description') {
          const [filename, declarationName] = args;

          const sourceFile = getSourceByFilename(filename);

          const declaration =
            declarationName === 'firstInterface'
              ? sourceFile.getInterfaces()[0]
              : sourceFile.getInterfaceOrThrow(declarationName);

          const allResults = [];

          const [jsdoc] = declaration.getJsDocs();
          if (jsdoc) {
            if (op === 'example') {
              const tags: Record<string, string> = {};
              for (const tag of jsdoc.getTags()) {
                tags[tag.getTagName()] = tag.getComment()?.trim() ?? '';
              }

              if (tags.example) {
                let parsed = tags.example;
                try {
                  parsed = JSON.stringify(JSON.parse(parsed), null, 4);
                } catch (error) {}

                allResults.push('```json');
                allResults.push(parsed);
                allResults.push('```\n');
              }
            } else if (op === 'description') {
              const description = jsdoc.getDescription().trim();
              allResults.push(description);
              allResults.push('\n');
            }
          }

          ignoreLine = true;

          updatedTemplateLines.push(line, ...allResults);
        } else if (op === 'installation') {
          updatedTemplateLines.push(line);
          if (packageName) {
            updatedTemplateLines.push(`\`npm i ${packageName}\`\n`);
            updatedTemplateLines.push(`or\n`);
            updatedTemplateLines.push(`\`yarn add ${packageName}\`\n`);
          }

          ignoreLine = true;
        } else if (op === 'break') {
          ignoreLine = false;
        }
      } else if (line[0] === '#') {
        ignoreLine = false;
      }

      if (!ignoreLine) {
        updatedTemplateLines.push(line);
      }
    }

    return ctx.file.text(filename, updatedTemplateLines.join('\n'));

    function getSourceByFilename(filename: string) {
      const resolvedFilename = Path.resolve(ctx.project.fullPath, filename);
      if (!sourceFiles[resolvedFilename]) {
        sourceFiles[resolvedFilename] = project.addSourceFileAtPath(
          resolvedFilename
        );
      }
      return sourceFiles[resolvedFilename];
    }
  }
);
