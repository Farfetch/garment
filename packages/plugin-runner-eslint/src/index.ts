import {
  defineOptionsFromJSONSchema,
  defineRunner,
  File
} from '@garment/runner';
import { CLIEngine } from 'eslint';

/**
 * Runs ESLint
 * @example {
    "runner": "eslint",
    "input": "{{projectFolder}}/src",
    "options": {
      "configFile": ".eslintconfig.json"
    }
  }
 */
interface EslintRunnerOptions {
  /**
   * The config file to load ESLint's configuration from.
   * @format path
   */
  configFile?: string;

  /**
   * Fix automatically linting errors.
   */
  fix?: boolean;
}

enum SeverityLevel {
  Warning = 1,
  Error = 2
}

export default defineRunner(
  defineOptionsFromJSONSchema<EslintRunnerOptions>(require('./schema.json')),
  async function(ctx) {
    const { logger, options, input, workspace, project } = ctx;
    const { configFile, outputDir, fix: fixFlag } = options;
    const fix = Boolean(outputDir) && fixFlag;

    /*
     * Updates the cwd to be the project
     * In this way the TS parser will only parse the project files
     */
    process.chdir(project.fullPath);

    logger.info(`Linting files with ESLint`);

    const cliEngineOpts: CLIEngine.Options = { fix, cwd: workspace.cwd };
    // if user specifies configFile should set useEslintrc to false
    if (configFile) {
      cliEngineOpts.configFile = configFile;
      cliEngineOpts.useEslintrc = false;
    }

    const cli = new CLIEngine(cliEngineOpts);

    ctx.input(files => {
      const fixes: File[] = [];

      files.forEach(file => {
        const { results } = cli.executeOnText(
          file.data.toString('utf8'),
          file.absolutePath
        );

        const warningResults = filterResultsBySeverity(
          results,
          SeverityLevel.Warning
        );
        const errorResults = filterResultsBySeverity(
          results,
          SeverityLevel.Error
        );
        const formatter = cli.getFormatter();
        const regexMatcher = /\n.*\u2716 .*\).*\n/;
        const fixableMessageMatcher = /[0-9]+ errors and [0-9]+ warnings .*/;
        if (errorResults.length) {
          const errorPrint = formatter(errorResults)
            .replace(regexMatcher, '')
            .replace(fixableMessageMatcher, '');
          logger.error(errorPrint);
        }
        if (warningResults.length) {
          logger.warn(
            formatter(warningResults)
              .replace(regexMatcher, '')
              .replace(fixableMessageMatcher, '')
          );
        }

        for (const { output } of results) {
          if (output && fix) {
            fixes.push(ctx.file.text(file.path, output));
          }
        }
      });

      return fixes;
    });
  }
);

function filterResultsBySeverity(
  results: CLIEngine.LintResult[],
  severity: SeverityLevel
) {
  const filtered: CLIEngine.LintResult[] = [];

  const count = (sevLevel: SeverityLevel, messagesSum: number) => {
    return severity === sevLevel ? messagesSum : 0;
  };

  results.forEach(result => {
    const filteredMessages = result.messages.filter(
      mess => mess.severity === severity
    );

    if (filteredMessages.length > 0) {
      filtered.push({
        ...result,
        messages: filteredMessages,
        errorCount: count(SeverityLevel.Error, filteredMessages.length),
        warningCount: count(SeverityLevel.Warning, filteredMessages.length),
        fixableErrorCount: result.fixableErrorCount,
        fixableWarningCount: 0
      });
    }
  });

  return filtered;
}
