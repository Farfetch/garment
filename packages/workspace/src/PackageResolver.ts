import * as Path from 'path';

export function requireDefault<T = any>(
  packageName: string,
  relativePath: string = ''
): T {
  const fullPath = Path.join(packageName, relativePath);

  const module = require(fullPath);
  if (typeof module.default === 'undefined') {
    throw new Error(`Module "${fullPath}" doesn't have default export`);
  }
  return module.default;
}

export function requireClass<T>(): T extends new (...args: any[]) => T
  ? T
  : never {
  return 0 as any;
}
