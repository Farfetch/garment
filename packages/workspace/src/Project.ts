import { Hash } from 'crypto';
import * as Mustache from 'mustache';
import * as Path from 'path';
import { ResolvedTaskConfig } from './Config';
import { NodePackage } from './NodePackage';

(Mustache as any).escape = (text: string) => text;

type Constructor<T> = new (...args: any[]) => T;

class ClassMap extends Map<Constructor<any>, any> {
  get<T>(key: Constructor<T>): T | undefined {
    return super.get(key);
  }
  set<T>(key: Constructor<T>, value: T) {
    return super.set(key, value);
  }
}

export class Project {
  readonly types = new ClassMap();

  readonly name: string;
  readonly path: string;
  readonly fullPath: string;
  readonly tasks: { [name: string]: ResolvedTaskConfig };
  readonly extendsSet: Set<string>;
  readonly dependecies: string[];

  nodePackage?: NodePackage;

  constructor(options: {
    readonly name: string;
    readonly path: string;
    readonly fullPath: string;
    readonly tasks: { [name: string]: ResolvedTaskConfig };
    readonly dependencies: string[];
    readonly extendsSet: Set<string>;
  }) {
    const { name, path, fullPath, tasks, dependencies, extendsSet } = options;
    this.name = name;
    this.path = path;
    this.fullPath = fullPath;
    this.tasks = tasks;
    this.extendsSet = extendsSet;
    this.dependecies = dependencies;
  }

  get aliases() {
    const aliases = new Set<string>();
    for (const [, type] of this.types) {
      if (typeof type.getAlias === 'function') {
        aliases.add(type.getAlias());
      }
    }
    return aliases;
  }

  resolvePath(...path: string[]) {
    return Path.resolve(this.fullPath, ...path);
  }

  resolvePathTemplate(pathTemplate: string = '') {
    return Mustache.render(pathTemplate, {
      projectDir: this.fullPath,
      projectName: this.name
    });
  }

  updateHash(hash: Hash) {
    hash.update(this.name).update(this.path);
    for (const type of this.types.values()) {
      if (typeof type.updateHash === 'function') {
        type.updateHash(hash);
      }
    }
  }
}
