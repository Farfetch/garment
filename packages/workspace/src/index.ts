export * from './Config';
export * from './PackageResolver';
export { Project } from './Project';
export { Workspace } from './Workspace';
