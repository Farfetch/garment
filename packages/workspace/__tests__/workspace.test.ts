import { getProjectsTasks } from '../src/Workspace';
import { Config } from '../src/Config';

describe('getProjectTasks', () => {
  const config: Config = {
    presets: {
      package: {
        tasks: {
          buildCSS: {
            runner: 'package-build-css'
          },
          build: [
            ':buildCSS',
            {
              runner: 'package-build'
            }
          ]
        }
      }
    },
    projects: {
      first: {
        path: 'projects/first',
        tasks: {
          clean: {
            runner: 'clean-first'
          },
          build: [
            ':clean',
            {
              runner: 'build'
            }
          ],
          prebuild: 'second:clean'
        }
      },
      second: {
        path: 'projects/second',
        extends: 'package',
        tasks: {
          clean: 'first:clean'
        }
      },
      third: {
        path: 'projects/third',
        tasks: {
          build: [
            {
              ref: 'first:build',
              next: [
                {
                  runner: 'build:foo'
                }
              ]
            }
          ]
        }
      },
      forth: {
        path: 'projects/forth',
        tasks: {
          prebuild: 'first:prebuild'
        }
      }
    }
  };

  test('resolves references to the same project task starting with ":"', () => {
    const taskDef = getProjectsTasks(config, 'first');

    expect(taskDef).toMatchSnapshot();
  });

  test('resolves tasks from preset and other project', () => {
    const taskDef = getProjectsTasks(config, 'second');

    expect(taskDef).toMatchSnapshot();
  });

  test('resolves tasks from preset and other project defined with ref', () => {
    const taskDef = getProjectsTasks(config, 'third');

    expect(taskDef).toMatchSnapshot();
  });

  test('throws ', () => {
    const taskDef = getProjectsTasks(config, 'forth');

    expect(taskDef).toMatchSnapshot();
  });
});
