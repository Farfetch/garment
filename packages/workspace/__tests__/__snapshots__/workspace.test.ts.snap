// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProjectTasks resolves references to the same project task starting with ":" 1`] = `
Object {
  "build": Object {
    "parallel": Array [
      Object {
        "id": "5504c2d4a8e70847a2bf687b93808ebd6f4f63e8",
        "runner": "clean-first",
      },
      Object {
        "id": "41fa59f6d7a356f5917ba7258dbb72fe262325f5",
        "runner": "build",
      },
    ],
  },
  "clean": Object {
    "id": "5504c2d4a8e70847a2bf687b93808ebd6f4f63e8",
    "runner": "clean-first",
  },
  "prebuild": Object {
    "id": "5504c2d4a8e70847a2bf687b93808ebd6f4f63e8",
    "runner": "clean-first",
  },
}
`;

exports[`getProjectTasks resolves tasks from preset and other project 1`] = `
Object {
  "build": Object {
    "parallel": Array [
      Object {
        "id": "674541ba19370bbd5ed06d0916a84ad61342175e",
        "runner": "package-build-css",
      },
      Object {
        "id": "bffe55daff7e64bbf2d4a4be94f5baa4f59ddb8c",
        "runner": "package-build",
      },
    ],
  },
  "buildCSS": Object {
    "id": "674541ba19370bbd5ed06d0916a84ad61342175e",
    "runner": "package-build-css",
  },
  "clean": Object {
    "id": "5504c2d4a8e70847a2bf687b93808ebd6f4f63e8",
    "runner": "clean-first",
  },
}
`;

exports[`getProjectTasks resolves tasks from preset and other project defined with ref 1`] = `
Object {
  "build": Object {
    "parallel": Array [
      Object {
        "next": Object {
          "parallel": Array [
            Object {
              "id": ">cdd80876be7565f1c86de3b32b057af4774aa759",
              "runner": "build:foo",
            },
          ],
        },
        "parallel": Object {
          "parallel": Array [
            Object {
              "id": "5504c2d4a8e70847a2bf687b93808ebd6f4f63e8",
              "runner": "clean-first",
            },
            Object {
              "id": "41fa59f6d7a356f5917ba7258dbb72fe262325f5",
              "runner": "build",
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`getProjectTasks throws  1`] = `
Object {
  "prebuild": Object {
    "id": "5504c2d4a8e70847a2bf687b93808ebd6f4f63e8",
    "runner": "clean-first",
  },
}
`;
